pkg/kubevirt/client.go

```diff
@@ -180,3 +213,126 @@ func (c *client) DeleteDataVolume(ctx context.Context, namespace string, name st
 	return c.cdiClient.CdiV1beta1().DataVolumes(namespace).Get(ctx, name, metav1.GetOptions{})
 }
+
+func (c *client) CreateVolumeSnapshot(ctx context.Context, namespace, name, claimName, snapshotClassName string) (*snapshotv1.VolumeSnapshot, error) {
+	snapshotClassName, err := c.getSnapshotClassNameFromVolumeName(ctx, namespace, claimName, snapshotClassName)
+	if err != nil {
+		return nil, err
+	}
+	snapshot := &snapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: snapshotv1.VolumeSnapshotSpec{
+			Source: snapshotv1.VolumeSnapshotSource{
+				PersistentVolumeClaimName: &claimName,
+			},
+		},
+	}
+	if snapshotClassName != "" {
+		snapshot.Spec.VolumeSnapshotClassName = &snapshotClassName
+	} else {
+		return nil, fmt.Errorf("no snapshot class name found for snapshot %s, %#v", name, snapshot)
+	}
+	klog.V(5).Infof("Creating snapshot %s with snapshot class %s, %#v", name, snapshotClassName, snapshot)
+	return c.snapClient.SnapshotV1().VolumeSnapshots(namespace).Create(ctx, snapshot, metav1.CreateOptions{})
+}
+
+func (c *client) getSnapshotClassNameFromVolumeName(ctx context.Context, namespace, claimName, snapshotClassName string) (string, error) {
+	volumeName, err := c.getVolumeNameFromClaimName(ctx, namespace, claimName)
+	if err != nil {
+		return "", err
+	} else if volumeName == "" {
+		return "", fmt.Errorf("persistent volume claim :%s not bound", claimName)
+	}
+	storageClassName, err := c.getStorageClassFromVolume(ctx, volumeName)
+	if err != nil {
+		return "", err
+	}
+	snapshotClass, err := c.getSnapshotClassFromStorageClass(ctx, storageClassName, snapshotClassName)
+	if err != nil {
+		return "", err
+	}
+	return snapshotClass.Name, nil
+}
+
+// Determine the name of the volume associated with the passed in claim name
+func (c *client) getVolumeNameFromClaimName(ctx context.Context, namespace, claimName string) (string, error) {
+	volumeClaim, err := c.kubernetesClient.CoreV1().PersistentVolumeClaims(namespace).Get(ctx, claimName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Error getting volume claim %s in namespace %s: %v", claimName, namespace, err)
+		return "", err
+	}
+	klog.V(5).Infof("found volumeClaim %#v", volumeClaim)
+	return volumeClaim.Spec.VolumeName, nil
+}
+
+// Determine the storage class from the volume
+func (c *client) getStorageClassFromVolume(ctx context.Context, volumeName string) (string, error) {
+	volume, err := c.kubernetesClient.CoreV1().PersistentVolumes().Get(ctx, volumeName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Error getting volume %s: %v", volumeName, err)
+		return "", err
+	}
+	return volume.Spec.StorageClassName, nil
+}
+
+// Get the associated snapshot class based on the storage class the following logic is used:
+// 1. If the snapshot class is provided AND the provisioner string matches, return that.
```

Comment:
what should be done if the snapshot class is provided, but it doesn't match the provisioner string?

Reply:
It should return an error indicating this. That lets the user know something is wrong.